pallets/parachain-system/src/lib.rs

```diff
@@ -364,6 +429,12 @@ pub mod pallet {
 				"ValidationData must be updated only once in a block",
 			);
 
+			// NOTE: the inherent data is expected to be unique, even if this block is built
+			// in the context of the same relay parent as the previous one. In particular,
+			// the inherent shouldn't contain messages that were already processed by any of the
+			// ancestors.
+			//
+			// This invariant should be upheld by the node-side.
```

Comment:
A comment for the benefit of other reviewers: the "node-side" here could also include the `ProvideInherent` implementation. That is probably the best way to ensure this in practice. Just trimming some MQCs.